Training24/intlist.c

```diff
@@ -0,0 +1,108 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------------------------------------
+// intlist.c
+// Program on A2 branch.
+// ------------------------------------------------------------------------------------------------
+
+#include <stdio.h>
+#include <malloc.h>
+#include "intlist.h"
+
+void Create (LinkList* list) {
+   list->head = NULL;
+   list->size = 0;
+}
+
+void DeleteList (LinkList* list) {
+   struct Node* current = list->head;  // Start from the head of the list
+   struct Node* nextNode;
+   while (current != NULL) {          //Traverse the list and free each node
+      nextNode = current->link;
+      free (current);                
+      current = nextNode;
+   }
+   list->head = NULL;                
+   list->size = 0;
+}
+
+void Add (LinkList* list, int value) {
+   if (list == NULL) return;
+   struct Node* newNode = (struct Node*)malloc (sizeof (struct Node));
+   if (newNode == NULL) return ERROR_MEMORY_ALLOCATION;
+   newNode->data = value;
+   newNode->link = NULL;
+   if (list->head == NULL) list->head = newNode;                   // If the list is empty the new node becomes the head
+   else {
+      struct Node* lastNode = list->head;
+      while (lastNode->link != NULL)  lastNode = lastNode->link;  // Traverse to the end of the list
+      lastNode->link = newNode;
+   }
+   list->size++;
+}
+
+void Insert (LinkList* list, int index, int value) {
+   if (index < 0 || index > list->size) return ERROR_OUT_OF_BOUNDS;
+   struct Node* newNode = (struct Node*)malloc (sizeof (struct Node));
+   newNode->data = value;
+   if (index == 0) {
+      newNode->link = list->head;
+      list->head = newNode;
+   }
+   else {
+      struct Node* currentNode = list->head;
+      for (int i = 0; i < index - 1; i++) currentNode = currentNode->link;    // Traverse to the node before the insertion
+      newNode->link = currentNode->link;
+      currentNode->link = newNode;
+   }
+   list->size++;
+}
+
+void RemoveAt (LinkList* list, int index) {
+   if (index < 0 || index > list->size) return ERROR_OUT_OF_BOUNDS;

```

Comment:
You should check if list-> head is null to prevent possible crashes. This is called d[efensive programming](https://en.wikipedia.org/wiki/Defensive_programming)